### PR TITLE
OtlpExporter: HttpClientFactory option

### LINKED ISSUE #2522: Does MeterProvider support OTLP HttpExporter?
# Question

I'm trying to emit metrics using OTLP HttpExporter (ConsoleExporter works) but nothing is getting sent. I wonder whether HttpExporter is supported for Metrics path?

```
  <ItemGroup>
    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.2.0-beta1" />
    <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.2.0-beta1" />
  </ItemGroup>
```

```
private static readonly Meter MyMeter = new Meter("ConsoleWithOTel.Meters", "1.0");
private static readonly Counter<long> IncomingRequestsCounter = MyMeter.CreateCounter<long>("IncomingRequests");

private static async Task Main(string[] args)
{
    using var meterProvider = Sdk.CreateMeterProviderBuilder()
        .AddMeter("ConsoleWithOTel.Meters")
        //.AddConsoleExporter(options => { options.MetricExportIntervalMilliseconds = 1000; })
        .AddOtlpExporter(opt =>
        {
            opt.MetricExportIntervalMilliseconds = 1000;
            opt.Endpoint = new Uri("https://bla.com/v1/metrics");
            opt.ExportProcessorType = ExportProcessorType.Simple; // tried Batch as well
            opt.Protocol = OtlpExportProtocol.HttpProtobuf; // Is this option supported?
        })
        .Build();

    IncomingRequestsCounter.Add(1, new KeyValuePair<string, object>("version", "v1"));
    IncomingRequestsCounter.Add(1, new KeyValuePair<string, object>("version", "v1"));
    IncomingRequestsCounter.Add(1, new KeyValuePair<string, object>("version", "v1"));
    IncomingRequestsCounter.Add(1, new KeyValuePair<string, object>("version", "v2"));
    IncomingRequestsCounter.Add(1, new KeyValuePair<string, object>("version", "v3"));
            
    Console.WriteLine($"{DateTimeOffset.UtcNow:O} Press any key to stop...");
    Console.ReadKey();
}
```

### PR DESCRIPTION
**Pull request by CodeBlanch**
Closes #2522

## Changes

* Adds the same `HttpClientFactory` option we have on `JaegerExporterOptions` & `ZipkinExporterOptions` to `OtlpExporterOptions`.
* Adds `IHttpClientFactory` integration when `IDeferredTracerProviderBuilder`/`IDeferredMeterProviderBuilder` is used.
* Adds support for the `HttpProtobuf` protocol in `OtlpMetricExporter`.

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [X] Unit tests

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using System.Threading;
+#endif
+using System.Threading.Tasks;
+using Google.Protobuf;
+using OtlpCollector = Opentelemetry.Proto.Collector.Metrics.V1;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.ExportClient
+{
+    /// <summary>Class for sending OTLP metrics export request over HTTP.</summary>
+    internal sealed class OtlpHttpMetricsExportClient : BaseOtlpHttpExportClient<OtlpCollector.ExportMetricsServiceRequest>
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs:33]: @alanwest I'm assuming we were just missing the client bits for this, but do you know if the collector actually supports metics over http?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs [hunk 2]
```diff
@@ -123,5 +134,38 @@ public OtlpExporterOptions()
         /// and Sum metrics.
         /// </summary>
         public AggregationTemporality AggregationTemporality { get; set; } = AggregationTemporality.Cumulative;
+
+        /// <summary>
+        /// Gets or sets the factory function called to create the <see
+        /// cref="HttpClient"/> instance that will be used at runtime to
+        /// transmit spans over HTTP. The returned instance will be reused for
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:141]:
```suggestion
        /// transmit telemetry over HTTP. The returned instance will be reused for
```

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:141]: Updated

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs:33]: Correct, just missing the bits. The collector does support metrics over HTTP (logs too).

Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs:33]: #2522

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpHttpMetricsExportClient.cs:33]: Cool thanks for confirming. I added a "Closes #2522" into the PR description. I can do logs too in a follow-up.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporterHelperExtensions.cs [hunk 3]
```diff
@@ -39,19 +41,61 @@ public static TracerProviderBuilder AddOtlpExporter(this TracerProviderBuilder b
...
+                    AddOtlpExporter(builder, sp.GetOptions<OtlpExporterOptions>(), configure, sp);
                 });
             }
 
-            return AddOtlpExporter(builder, new OtlpExporterOptions(), configure);
+            return AddOtlpExporter(builder, new OtlpExporterOptions(), configure, serviceProvider: null);
         }
 
-        private static TracerProviderBuilder AddOtlpExporter(TracerProviderBuilder builder, OtlpExporterOptions exporterOptions, Action<OtlpExporterOptions> configure = null)
+        internal static void BuildHttpClientFactory(IServiceProvider serviceProvider, OtlpExporterOptions options, string httpClientName)
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporterHelperExtensions.cs:51]: Maybe move this method to `OtlpExporterOptionsExtensions.cs`?

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporterHelperExtensions.cs]: Updated

**Review by alanwest — APPROVED** [review]
LGTM. Haven't thought it through, but this may actually be of use to the gRPC based exporters as well. At least for compilation that uses the Grpc.Net.Client library, you can set the underlying HttpClient to use. Allows for more sophisticated configuration of the channel. https://grpc.github.io/grpc/csharp-dotnet/api/Grpc.Net.Client.GrpcChannelOptions.html#Grpc_Net_Client_GrpcChannelOptions_HttpClient

**Review by cijothomas — APPROVED** [review]

**Pull request merged**